### ISSUE TITLE
Sort dask array

### ISSUE THREAD

**Issue by Pierre-Bartet** [dask/dask#4368]
Unless I missed it, there is no way to sort or argsort dask arrays, is it a current work in progress ?

**Comment by mrocklin** [comment]
I don't know of any current work on this, but it would be in scope if
anyone wants to start work on it.

**Comment by shoyer** [comment]
The easy version of this would be to sort along single dimensions only when they use a single chunk, and otherwise. The hard version would be to support sorting even in the multiple chunk case. I'd be glad to see either of these in dask.array.

**Comment by Pierre-Bartet** [comment]
Isn't it done somewhere in dask.dataframe.set_index ? I'll have a look.

**Comment by mrocklin** [comment]
Yes

**Comment by Pierre-Bartet** [comment]
For an intermediate step, I've made a something like this:

Make a graph -> compute it to obtain parameters p1, p2 -> Make another graph based on p1, p2 -> compute it to obtain desired result

But such a computation is not delayable, is there a way to delay a function which already compute a graph ?

**Comment by mrocklin** [comment]
> Isn't there a way to delay such a two (or many) stages computation so that from a user point of view set_index (or sort) is just a single lazy computation ?

Not if the structure of the computation depends on computed values.  When calling `dataframe.set_index` we need the divisions values for future dask.dataframe computations.  If you need to read data to compute these `divisions` values then things can't be entirely lazy.  If you're able to provide them manually then things will be entirely lazy.  

In the case of dask array you may not care about having that metadata upfront though, so this may not be an issue for you.

In short, if you're able to construct a task graph lazily then please do.  If you need to look at data to construct the task graph then no, there is no way to do that lazily.  (actually, there is, but it involves technology that we tend to avoid putting into dask array for complexity reasons).

**Comment by jakirkham** [comment]
cc @jrbourbeau (for awareness)

**Comment by sjperkins** [comment]
A k-way merge may be appropriate for the implementation: https://stackoverflow.com/questions/50907949/k-way-merge-sort-divided-on-multiple-hosts

**Comment by lucascolley** [comment]
Sorting would be good to have for SciPy! Of the ~575 failures so far for pushing `dask.array` through SciPy's test suite, ~150 are for missing `argsort`, and ~75 are for missing `sort`. So this is one of the biggest pieces of the array API standard that Dask is missing, with most of the other failures due to boolean indexing.

**Comment by lucascolley** [comment]
> The easy version of this would be to sort along single dimensions only when they use a single chunk, and otherwise error. The hard version would be to support sorting even in the multiple chunk case. I'd be glad to see either of these in dask.array.

The easy version sounds in scope for array-api-compat, but the hard version would probably have to be implemented in `dask.array` itself.

**Comment by crusaderky** [comment]
https://github.com/data-apis/array-api-compat/pull/239 adds sort and argsort by rechunking to a single chunk. That's better than nothing. It will work decently on some geometries and cause KilledWorker failures due to out-of-memory on others.

A better solution would be to implement proper sorting on top of shuffle, directly in daskbut that's an order of magnitude more complicated.

> Wouldn't the easy version bring no useful feature while misleading users into thinking that sorting is now possible in Dask?

It works fine IF squashing your array along a single chunk along the sort axis, while rechunking to auto size on all others, still results in reasonably-sized chunks.

**Comment by Pierre-Bartet** [comment]
Interesting to see that this is still open 6 years later!
I still find the solution to silently squash chunks very treacherous. Countless people will be bitten by such an unexpected behavior.  
As a user I haven't touched Dask in 6 years but given all the contributors can't we just all give a try and implement a clean merge sort?

**Comment by Pierre-Bartet** [comment]
I gave a look and it looks like it's still hard not in itself but because if I understand properly the existing shuffle (such as `set_index`) are still not lazy (which is what stopped me 6 years ago). Am I missing something?

**Comment by crusaderky** [comment]
Shuffles are a lazily-started MPI processes that run as worker and scheduler extensions.

**Comment by Pierre-Bartet** [comment]
Indeed, so could we do the same thing for `dask` arrays as what is done for `dask` dataframes `set_index`?  
Or am I missing something (for example `set_index` also squashing the chunk before sorting)?
Is `set_index` performing a k-way parallel merge sort?

**Comment by phofl** [comment]
sorting and set_index are **not** lazy, this is correct. plain shuffle is, but it doesn't guarantee order

**Comment by Pierre-Bartet** [comment]
Thanks I was questioning my own sanity!

**Comment by Pierre-Bartet** [comment]
I still don't get it, using `df.set_index("some_colulmn")` doesn't immediately trigger a computation, so it still look like at least somewhat lazy, yet I can't find a parallel merge sort happening somewhere.

**Comment by phofl** [comment]
The query optimizer made this kind of lazy for users, but not actually. The pre-compute is now triggered during optimization, i.e. if you run

``df.set_index(..).optimize()``. (you need more than a single partition for this)

The preemptive compute is a bit deferred but still happens before the actual compute is running

**Comment by Pierre-Bartet** [comment]
Ok thanks for explanation. Indeed now I understand why I was losing my sanity ^^  
Is this non lazy sorting also non parallel (does `dask` squashes the chunks before sorting)?

**Comment by phofl** [comment]
No, we aren't squashing anything here. We are calculating intervals of the column that we are sorting by, i.e.

x1 < x2 <... < xn

and then we are mapping the values into the new partitions based on what interval they are in

**Comment by Pierre-Bartet** [comment]
Nice, so then what is the limitation preventing a true lazy implementation?

**Comment by Pierre-Bartet** [comment]
Yes there are some aggregation then dispatching to be done but it is the case for many algorithms.

**Comment by pr38** [comment]
Hi,

Would it be in the cards to add sorting to dask array. I know it exists for Dataframes, I would pain me to cast to dataframe and back just to use sorting.
